### PR TITLE
0.9 Event Functionality

### PR DESCRIPTION
**Pull request by aumu1031**
Currently need to do manual DB update for mission_type = 6 (event) and mission rewards.

Reward column sample contents:
`[{"item_id":19,"chance":100,"quantity":3},{"item_id":20,"chance":50,"quantity":2}, {"item_id":21,"chance":5,"quantity":1}] `

### CODE REVIEW

File: classes/Mission.php [hunk 1]
```diff
@@ -305,6 +308,46 @@ public static function start($player, $mission_id): Mission {
 
         $mission = new Mission($mission_id, $player);
 
+        // TEMP Event Logic
+        global $system;
+        if ($mission->mission_type == Mission::TYPE_EVENT) {
+            if ($mission->rank == 2) {
+                echo "here";
```
Comment (levimeahan) [inline classes/Mission.php:315]: Is this debug/to be removed?

File: classes/Mission.php [hunk 2]
```diff
@@ -305,6 +308,46 @@ public static function start($player, $mission_id): Mission {
...
+                $valid = false;
+                foreach ($system->event_data['easy'] as $location) {
+                    if ($player->location->x == $location['x'] && $player->location->y == $location['y']) {
+                        echo $player->location->x;
```
Comment (levimeahan) [inline classes/Mission.php:319]: Is this debug/to be removed?

File: classes/Mission.php [hunk 3]
```diff
@@ -320,4 +363,47 @@ public static function maxMissionRank(int $player_rank): int {
...
+    public static function processRewards(Mission $mission, User $player, System $system): string {
+        if (count($mission->rewards) > 0) {
+            try {
+                $reward_text = "";
+                // load player inventory
+                $player->getInventory();
+
+                // get items from DB
+                $reward_keys = [];
+                foreach ($mission->rewards as $item) {
```
Comment (levimeahan) [inline classes/Mission.php:376]: If these are specifically item IDs, could we name this `$mission->reward_item_ids`?

File: classes/Mission.php [hunk 4]
```diff
@@ -320,4 +363,47 @@ public static function maxMissionRank(int $player_rank): int {
...
+                $reward_key_string = '(' . implode(',', $reward_keys) . ')';
+                /** @var Item[] $reward_items */
+                $reward_items = array();
+                $result = $system->db->query("SELECT * FROM `items` where `item_id` IN {$reward_key_string}");
+                while ($row = $system->db->fetch($result)) {
+                    $reward_items[$row['item_id']] = Item::fromDb($row);
+                }
+                // roll RNG, add to inventory
+                foreach ($mission->rewards as $item) {
+                    if (rand(0, 100) <= $item['chance']) {
```
Comment (levimeahan) [inline classes/Mission.php:388]:
```suggestion
                    if (mt_rand(0, 100) <= $item['chance']) {
```
Supposedly mt_rand is better than rand somehow. ALthough we could also use [random_int](https://www.php.net/manual/en/function.random-int.php)

File: classes/battle/BattleManager.php [hunk 5]
```diff
@@ -729,6 +729,8 @@ protected function applyAttack(BattleAttack $attack, Fighter $user, Fighter $tar
 
         $text = $attack->jutsu->battle_text;
         $attack_jutsu_color = BattleManager::getJutsuTextColor($attack->jutsu->jutsu_type);
+        $has_element = ($attack->jutsu->element != "none");
```
Comment (levimeahan) [inline classes/battle/BattleManager.php:732]:
```suggestion
        $has_element = ($attack->jutsu->element != Jutsu::ELEMENT_NONE);
```

File: db/migrations/20230628144927_mission_reward_migration.php [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+{
+    /**
+     * Migrate Up.
+     */
+    public function up(): void
+    {
+        // Modify missions table
+        $this->execute("
+            ALTER TABLE `missions` 
+                ADD `rewards` VARCHAR(500) NOT NULL DEFAULT '[]';
```
Comment (levimeahan) [inline db/migrations/20230628144927_mission_reward_migration.php:16]:
```suggestion
                ADD `rewards` VARCHAR(2000) NOT NULL DEFAULT '[]';
```
Lean really on the side of "way larger than we could ever fathom needing" with these if not using `TEXT`, it's tricky to track down when they're getting truncated

File: pages/event.php [hunk 7]
```diff
@@ -165,6 +165,47 @@ function event() {
...
     }
 
+    if (isset($_GET['exchange'])) {
+        try {
+            // Change the item ID values based on dev/prod
+            // These are placeholder exchange options
+            switch ($_GET['exchange']) {
+                case "red_yen":
+                    $player->getInventory();
+                    if (!$player->hasItem("19")) {
```
Comment (levimeahan) [inline pages/event.php:175]: Could we code the Item ID in event config somewhere? That will make it easier to swap for dev/prod

File: ui_components/src/home/Home.css [hunk 8]
```diff
@@ -962,27 +988,73 @@ body {
...
+        max-height: 55px;
+    }
+    .home_lantern_5::after {
+        content: '';
+        box-shadow: 0 0 45px 3px #f9f8dd, 0 0 20px 2px #c78b7b;
+        position: absolute;
+        width: 10px;
+        height: 40px;
+        left: 5px;
+        top: 5px;
```
Comment (levimeahan) [inline ui_components/src/home/Home.css:1088]: This part is the same for each of these, could we pull it into a `.home_lantern::after` class?

File: ui_components/src/home/News.js [hunk 9]
```diff
@@ -230,4 +234,11 @@ function NewsButtons({ loadNews, createPost, isAdmin }) {
             }
         </div>
     );
+}
+
+// Should probably be a global helper
+function clickOnEnter(event) {
+    if (event.key === "Enter") {
+        event.target.dispatchEvent(new MouseEvent('click', { bubbles: true }));
+    }
```
Comment (levimeahan) [inline ui_components/src/home/News.js:243]: This is a global helper now, we can remove from here

**Review by levimeahan — APPROVED** [review]
Looks good! Excited for this event

File: classes/Mission.php (hunk 1 above)
Comment (aumu1031) [reply classes/Mission.php:315]: Good catch, I'll remove this.

File: classes/Mission.php (hunk 2 above)
Comment (aumu1031) [reply classes/Mission.php:319]: Same as above.

File: ui_components/src/home/Home.css (hunk 8 above)
Comment (aumu1031) [reply ui_components/src/home/Home.css]: Resolved in commit.

File: ui_components/src/home/News.js (hunk 9 above)
Comment (aumu1031) [reply ui_components/src/home/News.js]: Resolved in commit.

File: classes/Mission.php (hunk 3 above)
Comment (aumu1031) [reply classes/Mission.php:374]: Resolved in commit.

**Pull request merged**